client/src/tests/ProductDetails/pdrender.test.js

```diff
@@ -1,40 +1,209 @@
+    render(<QuantitySelector selectedSku={selectedSku[0]} selectedStyle={selectedStyle} selectedQuantity={undefined} setSelectedQuantity={() => {}}/>)
+    
+    expect(screen.getByRole('combobox')).toBeTruthy();
+    expect(screen.getAllByRole('option').length).toBe(selectedSku[1].quantity + 1);
+    expect(screen.getByLabelText(`select quantity for ${selectedStyle?.name} sku ${selectedSku[0]}`)).toBeTruthy(); 
+
+  })
+
   it('should render cart actions', () => {
-    render(<ProductCartActions />);
+    const selectedStyle = ExampleProductStyles.results[0]; 
+    const inStockSkus = Object.entries(selectedStyle?.skus ?? {})?.filter(sku => (sku?.[1]?.quantity ?? 0) > 0);
+
+    render(<ProductCartActions selectedStyle={selectedStyle} />);
     expect(screen.getAllByRole('button').length).toBe(2);
-    //TODO: complete tests
+    expect(screen.getAllByRole('combobox').length).toBe(2);
+    const addToCartButton = screen.getByText('Add to Cart');
+    expect(addToCartButton).toBeTruthy(); 
+    //click without sku selected
+    fireEvent.click(addToCartButton); 
+    
+    //select a sku
+    const skuSelect = screen.getByText('Select Size');
+    expect(skuSelect).toBeTruthy(); 
+    fireEvent.change(skuSelect, {target: {value: inStockSkus[0][0] }});
+
+    fireEvent.click(addToCartButton); 
+
   });
 
-  it('should render product gallery', () => {
-    render(<ProductGallery />);
-    expect(screen.getByText('Carousel')).toBeTruthy();
-    expect(screen.getByText('Thumbnails')).toBeTruthy();
-    //TODO: complete tests
+  it('should render x product style selects & images', () => {
+    const setSelectedSyle = jest.fn(); 
+    render(<ProductStyleSelector selectedStyle={ExampleProductStyles.results[0]} productStyles={ExampleProductStyles} setSelectedStyle={setSelectedSyle}  />);
+    ExampleProductStyles.results.forEach((style) => {
+      expect(screen.getByLabelText(style.name)).toBeTruthy(); 
+      expect(screen.getByRole('img', {name: `style ${style.name} select preview`})).toBeTruthy();
+    });
+    const radioGroup = screen.getAllByRole('radio'); 
+    expect(radioGroup.length).toBeTruthy(); 
+    fireEvent.click(radioGroup[2]);
+    expect(setSelectedSyle).toHaveBeenCalledTimes(2); 
+  });
 
+  it('should render product price', () => {
+    render(<ProductPrice selectedStyle={ExampleProductStyles.results[0]} />);
+    expect(screen.getByText(`$${ExampleProductStyles.results[0].original_price}`)).toBeTruthy();
   });
 
-  it('should render style selector', () => {
-    render(<ProductStyleSelector />);
-    expect(screen.getByText('ProductStyleSelector')).toBeTruthy();
-    //TODO: complete tests
+  it('should render product sale price', () => {
+    render(<ProductPrice selectedStyle={ExampleProductStyles.results[2]} />);
+    expect(screen.getByText(`$${ExampleProductStyles.results[2].sale_price}`)).toBeTruthy();
+    expect(screen.getByText(`$${ExampleProductStyles.results[2].original_price}`)).toBeTruthy();
```

Comment:
once again i think these tests should be a little more specific/explicit, and you can also test to make sure that the sale price is red